client/web/src/search/useSourcegraphCodeIntelligence.ts

```diff
@@ -0,0 +1,71 @@
+import * as Monaco from 'monaco-editor'
+import { useEffect } from 'react'
+import { Observable } from 'rxjs'
+
+import { SearchPatternType } from '@sourcegraph/shared/src/graphql/schema'
+import { getDiagnostics } from '@sourcegraph/shared/src/search/query/diagnostics'
+import { getProviders } from '@sourcegraph/shared/src/search/query/providers'
+import { scanSearchQuery } from '@sourcegraph/shared/src/search/query/scanner'
+import { SearchSuggestion } from '@sourcegraph/shared/src/search/suggestions'
+
+export const SOURCEGRAPH_SEARCH = 'sourcegraphSearch' as const
+
+/**
+ * Adds code intelligence for the Sourcegraph search syntax to Monaco.
+ */
+export function useSourcegraphSearchCodeIntelligence(
```

Comment:
nit: since we're touching this code, how about we make this a little less verbose (feel free to do in follow up PR). `CodeIntelligence` has feels like an overloaded term.

```suggestion
export function useQueryIntelligence(
```

And correspondingly, this file could be `useQueryIntelligence` or similar.